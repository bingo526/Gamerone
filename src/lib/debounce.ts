// const debounce = (fn, time) => {
//   let timeout;

//   return function () {
//     const functionCall = () => fn.apply(this, arguments);

//     clearTimeout(timeout);
//     timeout = setTimeout(functionCall, time);
//   };
// };

// export default debounce;

const debounce = <F extends (...args: any[]) => any>(
  func: F,
  waitFor: number,
) => {
  let timeout: NodeJS.Timeout;

  return (...args: Parameters<F>): Promise<ReturnType<F>> =>
    new Promise((resolve) => {
      if (timeout) {
        clearTimeout(timeout);
      }

      timeout = setTimeout(() => resolve(func(...args)), waitFor);
    });
};

export default debounce;
