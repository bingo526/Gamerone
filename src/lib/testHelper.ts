export const waitForSec = (sec: number) =>
  new Promise((r) => setTimeout(r, sec));
