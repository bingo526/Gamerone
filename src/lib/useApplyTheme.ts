import { useRef, useLayoutEffect } from 'react';
import { useThemeState } from 'provider/theme';
import { G1_THEMES } from 'utils/constants';

export default function useApplyTheme() {
  const classRef = useRef<string[]>([]);
  const { theme, isSystemDefault, isDarkMode } = useThemeState();

  // Light/dark mode
  useLayoutEffect(() => {
    const htmlEl = document.querySelector('html');

    if (htmlEl) {
      classRef.current = htmlEl.classList
        .toString()
        .split(' ')
        .filter((s) => s);
      if (isSystemDefault) {
        htmlEl.classList.remove('prefersDarkMode', 'prefersLightMode');
      } else {
        htmlEl.classList.add(
          isDarkMode ? 'prefersDarkMode' : 'prefersLightMode',
        );
      }
    }

    // Clean up
    return () => {
      const htmlEl = document.querySelector('html');
      if (htmlEl && classRef.current) {
        const classes = htmlEl.classList
          .toString()
          .split(' ')
          .filter((s) => s);
        if (classes.length > 0) htmlEl.classList.remove(...classes);
        if (classRef.current.length > 0)
          htmlEl.classList.add(...classRef.current);
      }
    };
  }, [isSystemDefault, isDarkMode]);

  // Apply theme
  useLayoutEffect(() => {
    const bodyEl = document.querySelector('body');
    if (bodyEl) {
      bodyEl.classList.add(G1_THEMES[theme]);
    }

    // Clean up
    return () => {
      const bodyEl = document.querySelector('body');
      if (bodyEl) {
        bodyEl.classList.remove(G1_THEMES[theme]);
      }
    };
  }, [theme]);
}
