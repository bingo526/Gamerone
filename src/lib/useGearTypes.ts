import { useEffect, useState } from 'react';

import * as ReferenceDataApi from 'api/referenceData';
import { SelectOption } from 'components/common/Form/Input';
import { GearType } from 'interfaces';
import { Notify } from 'components/utility/Notify';

export default function useGearTypes() {
  const [gearTypeOptions, setGearTypeOptions] = useState<SelectOption[]>([]);

  useEffect(() => {
    const fetchGearTypes = async () => {
      try {
        const gearTypes = (await ReferenceDataApi.gearTypes()) as GearType[];

        setGearTypeOptions(
          gearTypes
            .sort((a, b) => a.ordinal - b.ordinal)
            .map(
              (type) =>
                ({
                  label: type.name,
                  value: type.id,
                } as SelectOption),
            ),
        );
      } catch (err) {
        Notify.warning('Gear ref data are missing');
      }
    };

    fetchGearTypes();
  }, []);

  return gearTypeOptions;
}
