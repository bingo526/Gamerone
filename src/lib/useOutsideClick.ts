import { useEffect, RefObject, useCallback } from 'react';

/**
 * Hook that handles outside click event of the passed refs
 *
 * @param refs array of refs
 * @param handler a handler function to be called when clicked outside
 */
export default function useOutsideClick(
  refs: Array<RefObject<HTMLElement> | undefined>,
  handler?: Function,
) {
  const handleClickOutside = useCallback(
    (event: any) => {
      if (!handler) return;

      // Clicked browser's scrollbar
      if (
        event.target === document.getElementsByTagName('html')[0] &&
        event.clientX >= document.documentElement.offsetWidth
      )
        return;

      let containedToAnyRefs = false;
      for (const rf of refs) {
        if (rf && rf.current && rf.current.contains(event.target)) {
          containedToAnyRefs = true;
          break;
        }
      }

      // Not contained to any given refs
      if (!containedToAnyRefs) {
        event.stopPropagation();

        handler();
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [handler],
  );

  useEffect(() => {
    // Bind the event listener
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [handleClickOutside]);
}
