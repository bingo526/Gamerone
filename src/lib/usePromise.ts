import { useState, useEffect } from 'react';
import { Nullable } from 'interfaces';

/**
 * An efficient custom hook that handles promise
 *
 * Caveat - Can't be used inside event handler, only useful inside component or hook
 * @param {Func} promiseCreator
 * @param {Array} deps  dependency array
 */
export default function usePromise<T>(
  promiseCreator: () => Promise<any>,
  deps: Array<any>,
) {
  const [loading, setLoading] = useState(false);
  const [resolved, setResolved] = useState<Nullable<T>>(null);
  const [error, setError] = useState<Nullable<Error>>(null);

  useEffect(() => {
    const process = async () => {
      setLoading(true);
      try {
        const resolved = await promiseCreator();
        setResolved(resolved);
      } catch (e) {
        setError(e);
      }
      setLoading(false);
    };
    process();
    return () => {
      setResolved(null);
      setError(null);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);

  return [loading, resolved, error] as const;
}
