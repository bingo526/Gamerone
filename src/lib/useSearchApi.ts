import { useState, useReducer, useEffect, useCallback } from 'react';
import { Nullable } from 'interfaces';

type SearchState<T> = {
  results: Array<T>;
  isSearching: boolean;
  isError: boolean;
};
type SearchAction<T> =
  | { type: 'SET_STATE'; payload: SearchState<T> }
  | { type: 'FETCH_INIT' }
  | { type: 'FETCH_SUCCESS'; payload: Array<T> }
  | { type: 'FETCH_FAILURE' };

const createDataFetchReducer = <T>() => (
  state: SearchState<T>,
  action: SearchAction<T>,
): SearchState<T> => {
  switch (action.type) {
    case 'SET_STATE':
      return {
        ...action.payload,
      };
    case 'FETCH_INIT':
      return {
        ...state,
        isSearching: true,
        isError: false,
      };
    case 'FETCH_SUCCESS':
      return {
        ...state,
        isSearching: false,
        isError: false,
        results: action.payload,
      };
    case 'FETCH_FAILURE':
      return {
        ...state,
        isSearching: false,
        isError: true,
      };
    default:
      throw new Error();
  }
};

export type SearchApi = (param?: Nullable<string>) => Promise<any>;

export default function useSearchApi<T>(
  apiAction: SearchApi,
  initialParam: Nullable<string> | undefined,
  initialData: Array<T>,
) {
  const [param, setParam] = useState(initialParam);
  const dataFetchReducer = createDataFetchReducer<T>();
  const [state, dispatch] = useReducer(dataFetchReducer, {
    isSearching: false,
    isError: false,
    results: initialData,
  });

  const setState = useCallback((state: SearchState<T>) => {
    dispatch({ type: 'SET_STATE', payload: state });
  }, []);

  useEffect(() => {
    let didCancel = false;
    const fetchData = async () => {
      dispatch({ type: 'FETCH_INIT' });
      try {
        const data = await apiAction(param);
        if (!didCancel) {
          dispatch({ type: 'FETCH_SUCCESS', payload: data });
        }
      } catch (error) {
        if (!didCancel) {
          dispatch({ type: 'FETCH_FAILURE' });
        }
      }
    };

    if (typeof param !== 'undefined' && param !== null) fetchData();

    return () => {
      didCancel = true;
    };
  }, [param, apiAction]);

  return [state, setParam, setState] as const;
}
