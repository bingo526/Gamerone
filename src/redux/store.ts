import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import { createBrowserHistory as createHistory } from 'history';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import thunk from 'redux-thunk';
import createSagaMiddleware from 'redux-saga';
import reducers from './reducers';
import rootSaga from './sagas';

const history = createHistory();
const sagaMiddleware = createSagaMiddleware();
const routeMiddleware = routerMiddleware(history);
const middlewares = [thunk, sagaMiddleware, routeMiddleware];

const composeEnhancers =
  (window['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__'] as typeof compose) || compose;

const store = createStore(
  combineReducers({
    ...reducers,
    router: connectRouter(history),
  }),
  composeEnhancers(applyMiddleware(...middlewares)),
);
sagaMiddleware.run(rootSaga);
export { store, history };
