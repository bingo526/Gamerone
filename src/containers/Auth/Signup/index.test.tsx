import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import { store } from 'redux/store';
import SignUpPage from './index';

import { waitForSec } from 'lib/testHelper';

/**
 * Mock input event to fill the form with email and password
 *
 * @param {String} email
 * @param {String} password
 * @return {Object}
 */
const renderSignUpPage = () => {
  const utils = render(
    <Provider store={store}>
      <MemoryRouter>
        <SignUpPage />
      </MemoryRouter>
    </Provider>,
  );

  return {
    ...utils,
  };
};

/**
 * Fill in the form and submit the form
 *
 * @param {Object} utils
 * @param {String} firstName
 * @param {String} lastName
 * @param {String} email
 * @param {String} username
 * @param {String} password
 */
const fillAndSubmitSignUpForm = async (
  utils,
  { firstName, lastName, email, username, password = 'password' },
) => {
  const { getByRole, getByLabelText } = utils;

  act(() => {
    fireEvent.change(getByLabelText(/firstname/i), {
      target: { value: firstName },
    });
    fireEvent.change(getByLabelText(/lastname/i), {
      target: { value: lastName },
    });
    fireEvent.change(getByLabelText(/username/i), {
      target: { value: username },
    });
    fireEvent.change(getByLabelText(/email/i), {
      target: { value: email },
    });
    fireEvent.change(getByLabelText(/password/i), {
      target: { value: password },
    });
    fireEvent.submit(getByRole('button', { name: 'Sign Up' }));
  });
};

test('checks validation for firstName, lastName and username fields', async () => {
  const { getByText, getByLabelText } = renderSignUpPage();
  const checkValidationText = async (elm, value, errorText) => {
    fireEvent.change(elm, { target: { value } });
    fireEvent.blur(elm);
    await waitForElement(() => getByText(errorText));
    expect(getByText(errorText)).toBeInTheDocument();
  };

  const firstNameElm = getByLabelText(/firstname/i);
  const usernameElm = getByLabelText(/username/i);
  const lastNameElm = getByLabelText(/lastname/i);

  // 1) Set 'firstName' empty
  await checkValidationText(firstNameElm, '', 'First name is required');

  // 2) Set a letter to 'firstName'
  await checkValidationText(
    firstNameElm,
    'x',
    'First name must be at least 2 letters long',
  );

  // 3) Set 'username' empty
  await checkValidationText(usernameElm, '', 'Username is required');

  // 4) Set 2 letters to 'username'
  await checkValidationText(
    usernameElm,
    'xx',
    'Username must be at least 3 letters long',
  );

  // 5) Set 'lastName' empty
  await checkValidationText(lastNameElm, '', 'Last name is required');

  // 6) Set a letter to 'lastName'
  await checkValidationText(
    lastNameElm,
    'y',
    'Last name must be at least 2 letters long',
  );
});

test('checks validation for email field', async () => {
  const { getByRole, getByText, getByLabelText } = renderSignUpPage();
  const submitForm = async () => {
    await act(async () => {
      fireEvent.submit(getByRole('button', { name: 'Sign Up' }));
    });
  };

  const emailElm = getByLabelText(/email/i);

  // 1) Set 'email' empty
  fireEvent.change(emailElm, { target: { value: '' } });
  await submitForm();
  expect(getByText('Email is required')).toBeInTheDocument();

  // 2) Set an incorrect email
  fireEvent.change(emailElm, { target: { value: 'xyz' } });
  await submitForm();
  expect(getByText('This is not correct email format')).toBeInTheDocument();

  // 3) Set an incorrect email
  fireEvent.change(emailElm, { target: { value: 'xyz@yy.' } });
  await submitForm();
  expect(getByText('This is not correct email format')).toBeInTheDocument();
});

test('checks duplicated email', async () => {
  const { getByText, getByLabelText, queryByLabelText } = renderSignUpPage();
  const duplicationMessage = 'This email is already registered';

  // mock out window.fetch for the test
  jest.spyOn(window, 'fetch').mockImplementationOnce(async () => {
    const error = new Error();
    error.code = 409;
    return Promise.reject(error);
  });

  const emailElm = getByLabelText(/email/i);

  // checks error message
  fireEvent.change(emailElm, { target: { value: 'duplicated@email.com' } });
  await waitForElement(() => getByText(duplicationMessage));
  expect(getByText(duplicationMessage)).toBeInTheDocument();

  // checks whether spinner is invisible
  expect(queryByLabelText(/spinner/i)).toBe(null);
});

test('checks duplicated username', async () => {
  const { getByText, getByLabelText, queryByLabelText } = renderSignUpPage();
  const duplicationMessage = 'This username is duplicated';

  // mock out window.fetch for the test
  jest.spyOn(window, 'fetch').mockImplementationOnce(async () => {
    const error = new Error();
    error.code = 409;
    return Promise.reject(error);
  });

  const usernameElm = getByLabelText(/username/i);

  // checks error message
  fireEvent.change(usernameElm, { target: { value: 'duplicated-username' } });
  await waitForElement(() => getByText(duplicationMessage));
  expect(getByText(duplicationMessage)).toBeInTheDocument();

  // checks whether spinner is invisible
  expect(queryByLabelText(/spinner/i)).toBe(null);
});

test('creates a new user and saves token inside localStorage', async () => {
  const utils = renderSignUpPage();
  const user = {
    email: 'xyz@domain.com',
    firstName: 'test',
    lastName: 'user',
    username: 'testuser',
    password: 'P@s2wordal',
    id: 6574,
  };
  const idToken = 'test_token';

  // mock out window.fetch for the test
  const spy = jest.spyOn(window, 'fetch').mockImplementation((url) => {
    let success = { status: 'success' };
    if (url.endsWith('/signup')) {
      success = { user, token: idToken };
    }
    return Promise.resolve({
      status: 200,
      json: () => success,
    });
  });

  await fillAndSubmitSignUpForm(utils, user);

  await waitForSec(1000);

  // check token with one from localStorage
  expect(utils.asFragment()).toMatchSnapshot();
  expect(localStorage.getItem('id_token')).toBe(idToken);

  // check user with one from the state
  expect(store.getState().Auth.user).toStrictEqual(user);

  spy.mockRestore();
});
