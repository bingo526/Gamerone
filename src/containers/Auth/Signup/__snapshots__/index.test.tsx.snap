// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a new user and saves token inside localStorage 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-lg-6 col-md-8 form-box"
      >
        <div
          class="col-md-12 form-key"
        >
          <i
            aria-hidden="true"
            class="fa fa-key"
          />
        </div>
        <div
          class="col-md-12 form-title"
        >
          Sign Up
        </div>
        <div
          class="col-md-12 form-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label
                class="form-control-label"
                for="firstName"
              >
                First Name
              </label>
              <input
                aria-label="firstName"
                class="form-control"
                name="firstName"
                type="text"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-control-label"
                for="lastName"
              >
                Last Name
              </label>
              <input
                aria-label="lastName"
                class="form-control"
                name="lastName"
                type="text"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-control-label"
                for="email"
              >
                Email
              </label>
              <input
                aria-label="email"
                class="form-control"
                name="email"
                type="text"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-control-label"
                for="username"
              >
                Username
              </label>
              <input
                aria-label="username"
                class="form-control"
                name="username"
                type="text"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-control-label"
                for="password"
              >
                Password
              </label>
              <input
                aria-label="password"
                class="form-control"
                name="password"
                type="password"
              />
            </div>
            <div
              class="col-lg-12 formbttm"
            >
              <div
                class="form-button"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="submit"
                >
                  Sign Up
                </button>
              </div>
            </div>
          </form>
        </div>
        <div
          class="col-md-12 form-footer"
        >
          <span />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
